Test/Microsoft.ApplicationInsights.Test/Shared/Extensibility/Implementation/Tracing/HeartbeatTests.cs

```diff
@@ -0,0 +1,415 @@
+        }
+
+        [TestMethod]
+        public void SentHeartbeatContainsExpectedDefaultFields()
+        {
+            using (var hbeat = new HealthHeartbeatProviderMock())
+            {
+                hbeat.Initialize(configuration: null);
+                hbeat.SimulateSend();
+                MetricTelemetry sentMsg = hbeat.sentMessages.First();
+                Assert.IsNotNull(sentMsg);
+
+                foreach (string field in HeartbeatDefaultPayload.DefaultFields)
+                {
+                    try
+                    {
+                        var fieldPayload = sentMsg.Properties.Single(a => string.Compare(a.Key, field) == 0);
+                        Assert.IsNotNull(fieldPayload);
+                        if (field.Equals(HeartbeatDefaultPayload.UpdatedFieldsPropertyKey, StringComparison.OrdinalIgnoreCase))
+                        {
+                            Assert.IsTrue(string.IsNullOrEmpty(fieldPayload.Value)); // updated fields should be empty for sdk default fields
+                        }
+                        else
+                        {
+                            Assert.IsFalse(string.IsNullOrEmpty(fieldPayload.Value)); // updated fields should not be empty
+                        }
+
+                    }
+                    catch (Exception)
+                    {
+                        Assert.Fail(string.Format(CultureInfo.CurrentCulture, "The default field '{0}' is not present exactly once in a sent message.", field));
+                    }
+                }
+            }
+        }
+
+        [TestMethod]
+        public void PayloadExtensionHandlesExtensionPayloadNameCollision()
+        {
+            using (var hbeat = new HeartbeatProvider())
+            {
+                hbeat.Initialize(configuration: null);
+
+                Assert.IsTrue(hbeat.AddHealthProperty("test01", "some test value", true));
+                Assert.IsFalse(hbeat.AddHealthProperty("test01", "some other test value", true));
+            }
+        }
+
+        [TestMethod]
+        public void CannotSetPayloadExtensionWithoutAddingItFirst()
+        {
+            using (var hbeat = new HeartbeatProvider())
+            {
+                hbeat.Initialize(configuration: null);
+
+                Assert.IsFalse(hbeat.SetHealthProperty("test01", "some other test value", true));
+                Assert.IsTrue(hbeat.AddHealthProperty("test01", "some test value", true));
+                Assert.IsTrue(hbeat.SetHealthProperty("test01", "some other test value", true));
+            }
+        }
+
+        [TestMethod]
+        public void CannotSetValueOfDefaultPayloadProperties()
+        {
+            using (var hbeat = new HeartbeatProvider())
+            {
+                hbeat.Initialize(configuration: null);
+
+                foreach (string key in HeartbeatDefaultPayload.DefaultFields)
+                {
+                    Assert.IsFalse(hbeat.SetHealthProperty(key, "test", true));
```

Comment:
minor-- I could not find the logic in code to prevent setting the value of DefaultPayload? Set is returning false because it's not finding the name in the extended items list.

Reply:
If you look at the HeartbeatProvider there is two lists of payload fields being maintained. One is the 'default' SDK provided properties, and the other is any properties that have been added by an extender. This is done to separate the two and avoid any issue with this exact issue - I don't think we want the user to have the ability to set default values of properties.

Reply:
I've changed how this works as of last week